aspnetcore/blazor/webassembly-native-dependencies.md

```diff
@@ -0,0 +1,159 @@
+
+## .NET WebAssembly build tools
+
+The .NET WebAssembly build tools are based on [Emscripten](https://emscripten.org/), a compiler toolchain for the web platform. To install the .NET WebAssembly build tools, use either of the following approaches:
+
+* Select the optional component in the Visual Studio installer.
+* Run `dotnet workload install wasm-tools` from an administrative command prompt.
+
+Add native dependencies to a Blazor WebAssembly app by adding `NativeFileReference` items in the app's project file. When the project is built, each `NativeFileReference` is passed to Emscripten by the .NET WebAssembly build tools so that they are compiled and linked into the runtime. Next, [`p/invoke`](/dotnet/standard/native-interop/pinvoke) into the native code from the app's .NET code.
+
+Generally, any portable native code can be used as a native dependency with Blazor WebAssembly. You can add native dependencies to C/C++ code or code previously compiled using Emscripten:
+
+* Object files (`.o`)
+* Archive files (`.a`)
+* Bitcode (`.bc`)
+* Standalone WebAssembly modules (`.wasm`)
+
+Prebuilt dependencies typically must be built using the same version of Emscripten used to build the .NET WebAssembly runtime.
+
+## Use native code
+
+Add a simple native C function to a Blazor WebAssembly app:
+
+1. Create a new Blazor WebAssembly project.
+1. Add the `WasmBuildNative` property to a property group in the app's project file (`.csproj`):
+
+   ```xml
+   <WasmBuildNative>true</WasmBuildNative>
+   ```
+
+   > [!NOTE]
+   > The `WasmBuildNative` property can also be set with the `-p|--property` option passed to the `dotnet run` command:
+   >
+   > ```dotnetcli
+   > dotnet run -p:WasmBuildNative=true
+   > ```
+
+1. Add a `Test.c` file to the project.
+1. Add a C function in `Test.c` for computing factorials:
+
+   ```c
+   int fact(int n)
+   {
+       if (n == 0) return 1;
+       return n * fact(n - 1);
+   }
+   ```
+
+1. Add a `NativeFileReference` for `Test.c` in the app's project file:
+
+   ```xml
+   <ItemGroup>
+     <NativeFileReference Include="Test.c" />
+   </ItemGroup>
+   ```
+
+1. In a Razor component (`.razor`), add a <xref:System.Runtime.InteropServices.DllImportAttribute> for the `fact` function in the generated `Test` library and call the `fact` method from .NET code in the component:
+
+   ```razor
+   @using System.Runtime.InteropServices
+
+   <p>@fact(3)</p>
+
+   @code {
+       [DllImport("Test")]
+       static extern int fact(int n);
+   }
+   ```
+
+When you build the app with the .NET WebAssembly build tools installed, the native C code is compiled and linked into the .NET WebAssembly runtime (`dotnet.wasm`). Compiling and linking may take a few minutes. After the app is built, run the app to see the rendered factorial value.
```

Comment:
In the final builds, debug relinking was sped up a lot and should now only take 5-10 seconds. So we could probably drop the statement that it "may take a few minutes".